pkg/agent/reset.go

```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		return "", "", fmt.Errorf("could not create osquery process to determine hardware UUID or serial: %w", err)
+	}
+
+	osqCtx, osqCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer osqCancel()
+
+	if sqlErr := osqProc.RunSql(osqCtx, []byte(query)); osqCtx.Err() != nil {
+		k.Slogger().Warn("query hardware UUID and serial context error", "err", osqCtx.Err())
+		return "", "", fmt.Errorf("querying hardware UUID and serial returned ctx error: %w", osqCtx.Err())
+	} else if sqlErr != nil {
+		return "", "", fmt.Errorf("querying hardware UUID and serial returned error: %w", sqlErr)
+	}
+
+	var resp []map[string]string
+	if err := json.Unmarshal(respBytes.Bytes(), &resp); err != nil {
+		return "", "", fmt.Errorf("unmarshalling response: %w", err)
+	}
+
+	if len(resp) < 1 {
+		return "", "", errors.New("no rows returned")
+	}
+
+	serial, ok := resp[0]["hardware_serial"]
+	if !ok {
+		return "", "", errors.New("hardware_serial missing from results")
+	}
+
+	hardwareUUID, ok := resp[0]["hardware_uuid"]
+	if !ok {
+		return "", "", errors.New("hardware_uuid missing from results")
+	}
+
+	return serial, hardwareUUID, nil
+}
+
+// valueChanged checks the knapsack for the given data and compares it to
+// currentValue. A value is considered changed only if the stored value was
+// previously set.
+func valueChanged(k types.Knapsack, currentValue string, dataKey []byte) bool {
+	storedValue, err := k.HostDataStore().Get(dataKey)
+	if err != nil {
+		k.Slogger().Error("could not get stored value", "err", err, "key", string(dataKey))
+		return false // assume no change
+	}
+
+	if storedValue == nil {
+		k.Slogger().Debug("value not previously stored, storing now", "key", string(dataKey))
+		if err := k.HostDataStore().Set(dataKey, []byte(currentValue)); err != nil {
+			k.Slogger().Error("could not store value", "err", err, "key", string(dataKey))
+		}
+		return false
+	}
+
+	if storedValue != nil && currentValue != string(storedValue) {
+		k.Slogger().Info("hardware-identifying value has changed", "key", string(dataKey))
+		return true
+	}
+
+	return false
+}
+
+// currentSecret retrieves the enrollment secret from either the knapsack or the filesystem,
+// depending on launcher configuration.
+func currentSecret(k types.Knapsack) (string, error) {
+	// Do we want to extract (and store) the tenant munemo, instead of storing the enrollment
+	// secret?
+	if k.EnrollSecret() != "" {
+		return k.EnrollSecret(), nil
+	}
+
+	if k.EnrollSecretPath() != "" {
+		content, err := os.ReadFile(k.EnrollSecretPath())
+		if err == nil {
+			return string(bytes.TrimSpace(content)), nil
+		}
+
+		return "", fmt.Errorf("could not read secret at enroll_secret_path %s: %w", k.EnrollSecretPath(), err)
+	}
+
+	return "", errors.New("enroll secret and secret path both unset")
+}
+
+// takeDatabaseBackup takes a backup of the current database and compresses it, storing
+// it in the root directory.
```

Comment:
Sure, I'm happy to do it that way. Maybe we could store that as a blob in the same `HostDataStore` bucket.

Stuff I could imagine us maybe caring about saving, beyond the items you mentioned:

* Configuration from `AgentFlagsStore` or `ConfigStore`?
* `ServerProvidedDataStore` has munemo, org id, device id, remote IP, and tombstone id -- maybe all of those?

I don't think we care about backing up the contents of `AutoupdateErrorsStore`, `ControlStore`, `InitialResultsStore`, `ResultLogsStore`, `OsqueryHistoryInstanceStore`, `SentNotificationsStore`, `StatusLogsStore`, `TokenStore` (since the tokens expire in a day anyway), or `ControlServerActionsStore`.

Reply:
I don't see much value in keeping the `AgentFlagsStore` or `ConfigStore`. I can't really imagine wanting any of that. I'm mostly thinking we should keep the things that help us diagnose what a device _was_ and correlate with our server logs.

The things you list in `ServerProvidedDataStore` seem right. I think we need an allowlist, since we might shove a lot of superfluous data in.

Reply:
I've updated to store the data discussed -- let me know how it looks to you